File: 卡码/代码随想录/Java/src/com/kurumi/codeThoughts/图论/dfs/LeetCode_0417.java
package com.kurumi.codeThoughts.图论.dfs;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 417. 太平洋大西洋水流问题
 */
public class LeetCode_0417 {
    static class Solution {
        int[][] dir = new int[][] { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };
        boolean[][][] visited;

        public void dfs(int[][] heights, int x, int y, int flag) {
            visited[flag][x][y] = true;
            for (int i = 0; i < 4; i++) {
                int nextX = x + dir[i][0];
                int nextY = y + dir[i][1];
                if (nextX < 0 || nextX > heights.length - 1 || nextY < 0 || nextY > heights[0].length - 1)
                    continue;
                if (visited[flag][nextX][nextY] || heights[nextX][nextY] < heights[x][y])
                    continue;
                dfs(heights, nextX, nextY, flag);
            }
        }

        public List<List<Integer>> pacificAtlantic(int[][] heights) {
            int n = heights.length;
            int m = heights[0].length;
            List<List<Integer>> res = new ArrayList<>();
            visited = new boolean[2][n][m];
            for (int i = 0; i < n; i++) {
                dfs(heights, i, 0, 0);
                dfs(heights, i, m - 1, 1);
            }
            for (int j = 0; j < m; j++) {
                dfs(heights, 0, j, 0);
                dfs(heights, n - 1, j, 1);
            }
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (visited[0][i][j] && visited[1][i][j])
                        res.add(Arrays.asList(i, j));
                }
            }
            return res;
        }
    }
}

File: 卡码/代码随想录/Go/图论/bfs/LeetCode_0417.go
// 417. 太平洋大西洋水流问题
package main

func pacificAtlantic(heights [][]int) [][]int {
	dir := [2][4]int{[4]int{0, 1, 0, -1}, [4]int{1, 0, -1, 0}}
	n, m := len(heights), len(heights[0])
	visited := make([][][]bool, 2)
	for i := 0; i < 2; i++ {
		visited[i] = make([][]bool, n)
		for j := 0; j < n; j++ {
			visited[i][j] = make([]bool, m)
		}
	}
	q := make([][3]int, 0)
	var bfs func()
	bfs = func() {
		for len(q) > 0 {
			v := q[0]
			q = q[1:]
			for i := 0; i < 4; i++ {
				nextX, nextY := v[1]+dir[0][i], v[2]+dir[1][i]
				if nextX < 0 || nextX > n-1 || nextY < 0 || nextY > m-1 {
					continue
				}
				if visited[v[0]][nextX][nextY] || heights[nextX][nextY] < heights[v[1]][v[2]] {
					continue
				}
				q = append(q, [3]int{v[0], nextX, nextY})
				visited[v[0]][nextX][nextY] = true
			}
		}
	}
	for i := 0; i < n; i++ {
		visited[0][i][0] = true
		visited[1][i][m-1] = true
		q = append(q, [3]int{0, i, 0})
		q = append(q, [3]int{1, i, m - 1})
	}
	for j := 0; j < m; j++ {
		visited[0][0][j] = true
		visited[1][n-1][j] = true
		q = append(q, [3]int{0, 0, j})
		q = append(q, [3]int{1, n - 1, j})
	}
	bfs()
	res := make([][]int, 0)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if visited[0][i][j] && visited[1][i][j] {
				res = append(res, []int{i, j})
			}
		}
	}
	return res
}
